GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    /// Implements the NamedPipe protocol as described in NamedPipeServer.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        /// <summary>
+        /// Read a message from the stream.
+        /// </summary>
+        /// <returns>The message read from the stream, or null if the end of the input stream has been reached. </returns>
+        public string ReadMessage()
+        {
+            int bytesRead = this.Read();
+            if (bytesRead == 0)
+            {
+                // The end of the input stream has been reach - return null to indicate this.
+                return null;
+            }
+
+            // If we have read in the entire message in the first read (mainline scenario),
+            // then just process the data directly from the buffer.
+            if (this.buffer[bytesRead - 1] == TerminatorByte)
+            {
+                return Encoding.UTF8.GetString(this.buffer, 0, bytesRead);
+            }
+
+            // We need to process multiple chunks - collect data from multiple chunks
+            // into a single list
+            List<byte> bytes = new List<byte>(BufferLength * 2);
+
+            while (true)
+            {
+                bool encounteredTerminatorByte = this.buffer[bytesRead - 1] == TerminatorByte;
+                int lengthToCopy = encounteredTerminatorByte ? bytesRead - 1 : bytesRead;
+
+                bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, lengthToCopy));
+                if (encounteredTerminatorByte)
+                {
+                    break;
+                }
+
+                bytesRead = this.Read();
+
+                if (bytesRead == 0)
+                {
+                    // We have read a partial message (the last byte received does not indicate that
+                    // this was the end of the message), but the stream has been closed. Throw an exception
+                    // and let upper layer deal with this condition.
+
+                    throw new IOException("Incomplete message read from stream. Stream closed before message terminating byte.");

```

Comment:
I would throw a `BrokenPipeException` here

Reply:
I can do that - I was trying to decide between those two exceptions. Currently, `NamedPipeClient` and `NamedPipeServer` both handle `IOException` , which is why I chose this exception. If I switch to `BrokenPipeException`, I will also update the handling code in `NamedPipeServer` to handle `BrokenPipeExceptions` in the same manner as it currently does for `BrokenPipeExceptions`.

I can go with either approach - do you have a preference?

Reply:
I'm guessing we were handling IOExceptions before because that's what the built in reader/writer were throwing. I would update this to throw/catch our specific exception type, and remove the handler for IOException unless there's still a valid reason for it.

Reply:
Yeah - the underlying Stream class can still throw `IOException`. We could have the `NamedPipeStreamReader` and `NamedPipeStreamWriter` class catch and wrap all underlying `IOException`s as `BrokenPipeException`s.

Would you be OK with keeping this as an IOException for now, which would allow us to keep all existing error handling in place? I think it would be easier if we extracted that change into its isolated change, so we can see the update to the exception type (and all handlers) by itself?